### PR TITLE
Introduce `StorageFactorySwitch`

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**

### CODE REVIEW

File: server/src/main/java/org/spine3/server/storage/StorageFactorySwitch.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+/**
+ * The {@code Supplier} of {@code StorageFactory} that takes into account
+ * if the code runs under tests.
+ *
+ * <h2>Test mode</h2>
+ * <p>Under tests  this class returns {@link InMemoryStorageFactory} if
+ * a {@code Supplier} for tests was not set via {@link #init(Supplier, Supplier)}.
+ *
+ * <h2>Production mode</h2>
+ * <p>In production mode this class obtains the instance obtained
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/storage/StorageFactorySwitch.java]: `...obtains the instance obtained...` looks weird.

I think `obtains the instance provided by the` will do the job.

File: server/src/main/java/org/spine3/server/storage/StorageFactorySwitch.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+ * by the {@code Supplier} passed via {@link #init(Supplier, Supplier)}.
+ * If the production {@code Supplier} was not initialized {@code IllegalStateException}
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/storage/StorageFactorySwitch.java]: I think, a comma after the `not initialized` would improve the readability.

File: server/src/main/java/org/spine3/util/Environment.java [hunk 3]
```diff
@@ -20,30 +20,80 @@
...
+     * set this property to {@code true} before running tests.
+     */
+    public static final String ENV_KEY_TESTS = "org.spine3.tests";
 
     /** The key of the Google AppEngine runtime version system property. */
-    private static final String APP_ENGINE_RUNTIME_VERSION_KEY = "com.google.appengine.runtime.version";
+    @VisibleForTesting
+    static final String ENV_KEY_APP_ENGINE_RUNTIME_VERSION = "com.google.appengine.runtime.version";
+
+    /** If set contains the version of AppEngine obtained from the system property. */
```
Comment (armiol) [inline server/src/main/java/org/spine3/util/Environment.java]: Let's add a comma after `If set`.

File: server/src/main/java/org/spine3/util/Environment.java [hunk 4]
```diff
@@ -57,9 +107,76 @@ public boolean isAppEngine() {
...
+            this.tests = true;
+            return true;
+        }
+
+        this.tests = false;
+        return false;
+    }
+
+    /**
+     * Turns the environment into tests mode.
```
Comment (armiol) [inline server/src/main/java/org/spine3/util/Environment.java]: Perhaps, `Switches ...`?
Or maybe `Turns on the test mode for the environment.`?

File: server/src/test/java/org/spine3/server/storage/StorageFactorySwitchShould.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    @After
+    public void cleanUp() throws NoSuchFieldException, IllegalAccessException {
+        cleanSwitch();
+        Environment.getInstance().reset();
+    }
+
+    /**
+     * Cleans the `storageFactorySwitch` instance by nullifying fields.
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/storage/StorageFactorySwitchShould.java]: I think, `clear` will be more suitable as an action verb for the intention like this.

Another option to consider is `restore`, if you want to emphasise the returning to the original state.

File: server/src/test/java/org/spine3/server/storage/StorageFactorySwitchShould.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        Environment.getInstance().reset();
+        // Pretend that we are not under tests for the `Environment`.
+        Environment.getInstance().setToProduction();
+
+        storageFactorySwitch.get();
+    }
+
+    @Test
+    public void cache_instance_of_StorageFactory_in_testing() {
+        Supplier<StorageFactory> testingSupplier = spy(inMemorySupplier);
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/storage/StorageFactorySwitchShould.java]: `final`?

File: server/src/test/java/org/spine3/server/storage/StorageFactorySwitchShould.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+        storageFactorySwitch.get();
+        storageFactorySwitch.get();
+
+        verify(testingSupplier, times(1)).get();
+    }
+
+    @Test
+    public void cache_instance_of_StorageFactory_in_production() {
+        Supplier<StorageFactory> productionSupplier = spy(inMemorySupplier);
```
Comment (armiol) [inline server/src/test/java/org/spine3/server/storage/StorageFactorySwitchShould.java]: `final`?

**Comment by alexander-yevsyukov** [comment]
@armiol PTAL

**Pull request merged**